db/view/view.cc

```diff
@@ -3576,5 +3582,453 @@ sstring build_status_to_sstring(build_status status) {
+    for (auto& [shard, shard_tasks]: new_tasks) {
+        if (!is_shard_free(shard)) {
+            vbw_logger.warn("Tried to start new view building tasks on shard {} but the shard is busy", shard);
+            throw std::runtime_error(fmt::format("Shard {} is already busy", shard));
+        }
+
+        auto tasks = shard_tasks | std::views::transform([] (const service::view_building::view_building_task& t) {
+            return std::make_pair(t.id, t);
+        }) | std::ranges::to<std::unordered_map>();
+        auto batch = seastar::make_shared<view_building_worker::batch>(*this, tasks, shard_tasks.front().base_id, shard_tasks.front().replica);
+        for (auto& [id, _]: tasks) {
+            _state.tasks_map.insert({id, batch});
+        }
+        vbw_logger.debug("Started tasks {} for base table {} on shard {}", tasks | std::views::keys, shard_tasks.front().base_id, shard_tasks.front().replica.shard);
+        
+        co_await coroutine::maybe_yield();
+    }
+}
+
+bool view_building_worker::is_shard_free(shard_id shard) {
+    return !std::ranges::any_of(_state.tasks_map, [&shard] (auto& task_entry) {
+        return task_entry.second->replica.shard == shard;
+    });
+}
+
+void view_building_worker::init_messaging_service() {
+    ser::view_rpc_verbs::register_work_on_view_building_tasks(&_messaging, [this] (std::vector<utils::UUID> ids) -> future<std::vector<service::view_building::view_task_result>> {
+        return container().invoke_on(0, [ids = std::move(ids)] (view_building_worker& vbw) -> future<std::vector<service::view_building::view_task_result>> {
+            return vbw.get_tasks_results(std::move(ids));
+        });
+    });
+}
+
+future<> view_building_worker::uninit_messaging_service() {
+    return ser::view_rpc_verbs::unregister(&_messaging);
+}
+
+future<std::vector<service::view_building::view_task_result>> view_building_worker::get_tasks_results(std::vector<utils::UUID> ids) {
+    vbw_logger.debug("Got request for results of tasks: {}", ids);
+
+    // We assume that all tasks in `ids` are executed in the same batch.
+    // Group0_guard works as semaphore here and keeps view building state locked.
+    auto wait_until_batch_is_finished = [this, &ids] () -> future<service::group0_guard> {
+        auto id = ids.front();
+        
+        while (!_as.abort_requested()) {
+            auto guard = co_await _group0_client.start_operation(_as);
+
+            if (_state.tasks_map.contains(id)) {
+#ifdef SEASTAR_DEBUG
+                // validate that all tasks are in the same batch
+                for (auto& tid: ids) {
+                    SCYLLA_ASSERT(_state.tasks_map[id]->tasks.contains(tid));
+                }
+#endif
+                service::release_guard(std::move(guard));
+                vbw_logger.debug("Batch with task {} is not finish yet, waiting on CV", id);
+                _vb_state_machine.event.broadcast(); // wake up the worker so it'll check if the batch was finished
```

Comment:
Why is this needed? Shouldn't the worker track how many remaining tasks are there and notify the `batch_done_cv` at the right time itself?

Reply:
removed